docarray/array/storage/memory/getsetdel.py

```diff
@@ -26,7 +26,7 @@ def _del_docs_by_slice(self, _slice: slice):
 
     def _del_doc_by_id(self, _id: str):
         del self._data[self._id2offset[_id]]
-        self._id2offset.pop(_id)
```

Comment:
calling `_rebuild_id2offset` is expensive. in general every inconsistency problem can be solved by calling this method, but we choose to avoid that for reducing the overhead.

What is exact problem of the current `_id2offset.pop` logic?

Reply:
Line 28 when we remove the Document from the data the indices from all of the following document shift by one which is not reflected by self._id2offset.pop. Technically instead of rebuilding the offset we could decrement the indices of only the documents that are after the one that just got deleted. But unless we are missing something, that wouldn't be a big gain (still O(n)).5

Reply:
Currently that is what happening :

```python
da = DocumentArray([Document(id=f"doc{i}") for i in range(4)])
del da["doc0"]
del da["doc1"]
print(da[:,"id"]) >>> ['doc1', 'doc3']
```

when calling del da["doc1"] it is the "doc2" which is deleted because the id2offset did not shift the indices

Reply:
you are right, but this suggest `insert` logic is also wrong

Reply:
`docarray/array/storage/memory/seqlike.py:L16`

Reply:
Yep you are right, it is fix now in the last commit